### PR TITLE
apps/shell/tash : Add command autocompletion support

### PR DESCRIPTION
**Pull request by YiHyunJin**
Press tab while typing commands to auto complete.
Pressing tab twice while typing a command will show
you the commands you can use.

TASH>>hel
help             hello

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [83c4c934df0b76548ab1160965eea3c6a270917d] - Code Rule Check Result:
apps/shell/tash_command.c:291: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
apps/shell/tash_command.c:300: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [83c4c934df0b76548ab1160965eea3c6a270917d] - Code Rule Check (C++) OK.

File: apps/shell/tash_command.c [hunk 1]
```diff
@@ -269,6 +269,50 @@ int tash_execute_cmd(char **args, int argc)
 	return 0;					/* Need to pass the appropriate error later */
 }
 
+bool tash_tab_cmd(char *cmd, int *pos, bool pressed_tab)
+{
+	int cmd_idx;
+	int found_cnt = 0;
+	int idx = -1;
+	/* lock mutex */
+	pthread_mutex_lock(&tash_cmds_info.tmutex);
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c]: How about moving this below the declaration of search array?

**Comment by seinfra** [comment]
Target : [13878707b02704a88a5f58515fa2e0ba515ddcb9] - Code Rule Check Result:
apps/shell/tash_command.c:291: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
apps/shell/tash_command.c:301: ERROR: [SPC_M_KWD] space required before the open parenthesis '('

**Comment by seinfra** [comment]
Target : [13878707b02704a88a5f58515fa2e0ba515ddcb9] - Code Rule Check (C++) OK.

File: apps/shell/tash_command.c [hunk 2]
```diff
@@ -269,6 +269,51 @@ int tash_execute_cmd(char **args, int argc)
...
+			search[cmd_idx] = true;
+		}
+	}
+
+	if (found_cnt == 1 && pressed_tab == false) {
+		while(tash_cmds_info.cmd[idx].str[*pos] != 0) {
+			cmd[*pos] = tash_cmds_info.cmd[idx].str[*pos];
+			(*pos)++;
+		}
+		pthread_mutex_unlock(&tash_cmds_info.tmutex);
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c]: Please merge these unlock and return mechanism.

File: apps/shell/tash_command.c [hunk 3]
```diff
@@ -269,6 +269,51 @@ int tash_execute_cmd(char **args, int argc)
...
+	for (cmd_idx = 0; cmd_idx < tash_cmds_info.count; cmd_idx++) {
+		if (strncmp(cmd, tash_cmds_info.cmd[cmd_idx].str, *pos) == 0) {
+			idx = cmd_idx;
+			found_cnt++;
+			search[cmd_idx] = true;
+		}
+	}
+
+	if (found_cnt == 1 && pressed_tab == false) {
+		while(tash_cmds_info.cmd[idx].str[*pos] != 0) {
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c]: Please check the coding rule.

File: apps/shell/tash_command.c [hunk 4]
```diff
@@ -269,6 +269,51 @@ int tash_execute_cmd(char **args, int argc)
...
+			cmd[*pos] = tash_cmds_info.cmd[idx].str[*pos];
+			(*pos)++;
+		}
+		pthread_mutex_unlock(&tash_cmds_info.tmutex);
+		return true;
+	} else if (found_cnt > 1 && pressed_tab == true) {
+		int cnt = 0;
+		printf("\n");
+		for (cmd_idx = 0; cmd_idx < tash_cmds_info.count; cmd_idx++) {
+			if(search[cmd_idx] == true) {
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c]: Please check the coding rule.

File: apps/shell/tash_command.c (hunk 1 above)
Comment (YiHyunJin) [reply apps/shell/tash_command.c]: done.

**Comment by seinfra** [comment]
Target : [ab07e37dd8c34d9dbac5cbdb5c53c812bf3cc735] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ab07e37dd8c34d9dbac5cbdb5c53c812bf3cc735] - Code Rule Check (C++) OK.

File: apps/shell/tash_command.c [hunk 5]
```diff
@@ -269,6 +269,50 @@ int tash_execute_cmd(char **args, int argc)
 	return 0;					/* Need to pass the appropriate error later */
 }
 
+bool tash_tab_cmd(char *cmd, int *pos, bool pressed_tab)
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c]: Please add description about this function.

**Comment by seinfra** [comment]
Target : [7d53218992ade2ea69625b1a38483384d9c4e353] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7d53218992ade2ea69625b1a38483384d9c4e353] - Code Rule Check (C++) OK.

File: apps/shell/tash_command.c [hunk 6]
```diff
@@ -269,6 +269,68 @@ int tash_execute_cmd(char **args, int argc)
...
+						   false - tab is pressed once
+ * @retval true Success
+ * @retval false Failure
+ *
+ * Example Usage:
+ * @code
+ *
+ * @endcode
+ */
+bool tash_tab_cmd(char *cmd, int *pos, bool pressed_tab)
```
Comment (jeongchanKim) [inline apps/shell/tash_command.c]: I think this function can be static. If then, we don't need to declare prototype in tash_internal.h

**Comment by seinfra** [comment]
Target : [6bb0b4c99c459fccdb21c75eb8403e4ffe445bab] - Code Rule Check Result:
apps/shell/tash_command.c:278: ERROR: [IDT_M_TAB] please, no space before tabs

**Comment by seinfra** [comment]
Target : [6bb0b4c99c459fccdb21c75eb8403e4ffe445bab] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4f81baf04ce03ad21c25576cee8a8e8060788633] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4f81baf04ce03ad21c25576cee8a8e8060788633] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6d15cd75b27cc13619f13d658d86361c9d797e33] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [6d15cd75b27cc13619f13d658d86361c9d797e33] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [aafca30acbe1e40a4b2c66cec2275d4f46a23989] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [aafca30acbe1e40a4b2c66cec2275d4f46a23989] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3d3388f4c3fa26c357759458165e446dc72fa640] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [3d3388f4c3fa26c357759458165e446dc72fa640] - Code Rule Check (C++) OK.

File: apps/shell/tash_command.c [hunk 7]
```diff
@@ -269,6 +269,58 @@ int tash_execute_cmd(char **args, int argc)
...
 
+/** @name tash_do_autocomplete
+ * @brief API to TASH tab commands
+ * @ingroup tash
+ * @param[in] cmd - command's first string
+ * @param[in] pos - length of command's first string
+ * @param[in] double_click : true - tab is pressed more than once
+ *                           false - tab is pressed once
+ */
+bool tash_do_autocomplete(char *cmd, int *pos, bool double_click)
```
Comment (sunghan-chang) [inline apps/shell/tash_command.c]: Let's rename the double_click.

**Comment by seinfra** [comment]
Target : [5cb134823c54a96d10956507e5bdc2ba5389ebc3] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5cb134823c54a96d10956507e5bdc2ba5389ebc3] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fa46e0a9ac6036b3da3bb60e5edf7b3fb8bf7469] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fa46e0a9ac6036b3da3bb60e5edf7b3fb8bf7469] - Code Rule Check (C++) OK.

**Pull request merged**